File: backend/src/agents/agenticChat.ts
/**
 * Agentic AI Chat Agent with LLM-Based Function Calling
 * LLM decides which function to call (calendar event, policy docs, or direct answer)
 */

import { pool } from '../db';
import { generateChatCompletion, LLMMessage } from '../clients/llm';
import * as fs from 'fs';
import * as path from 'path';

// Agent state interface
export interface AgentState {
  messages: Array<{ role: 'user' | 'assistant'; content: string; timestamp: Date }>;
  sessionId: string;
  context: {
    inboxStats?: string;
    tasksStats?: string;
    lastActivity: Date;
  };
  tools: {
    calendarEventCreated?: any;
  };
}

// Session storage
const agentSessions = new Map<string, AgentState>();

// Clean up inactive sessions (older than 1 hour)
setInterval(() => {
  const now = Date.now();
  const oneHour = 60 * 60 * 1000;
  
  for (const [sessionId, state] of agentSessions.entries()) {
    if (now - state.context.lastActivity.getTime() > oneHour) {
      agentSessions.delete(sessionId);
      console.log(`[Agent] Cleared inactive session: ${sessionId}`);
    }
  }
}, 5 * 60 * 1000); // Every 5 minutes

// Tool: Create calendar event
async function createCalendarEvent(text: string): Promise<{ success: boolean; event?: any; message: string }> {
  try {
    const port = process.env.PORT || 3001;
    const calendarParseUrl = `http://localhost:${port}/api/calendar/parse`;
    
    console.log(`[Agent] Creating calendar event via: ${calendarParseUrl}`);
    console.log(`[Agent] Request text: ${text}`);
    
    const response = await fetch(calendarParseUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ text }),
    });

    console.log(`[Agent] Calendar API response status: ${response.status}`);

    if (response.ok) {
      const data = await response.json() as { event?: any; parsed?: any };
      console.log(`[Agent] Calendar API response:`, JSON.stringify(data).substring(0, 200));
      
      if (data.event) {
        console.log(`[Agent] ✅ Calendar event created: ${data.event.title}`);
        return {
          success: true,
          event: data.event,
          message: `Calendar event "${data.event.title}" created successfully.`,
        };
      } else {
        console.warn(`[Agent] Calendar API returned OK but no event in response`);
      }
    } else {
      const errorText = await response.text();
      console.error(`[Agent] Calendar API error (${response.status}): ${errorText}`);
    }
    
    return { success: false, message: 'Failed to create calendar event.' };
  } catch (error: any) {
    console.error('[Agent] Calendar event creation error:', error);
    console.error('[Agent] Error stack:', error.stack);
    return { success: false, message: `Error: ${error.message}` };
  }
}

// Tool: Get policy document
function getPolicyDocument(): string {
  const possiblePaths = [
    path.join(__dirname, '../../policies/company-policy.md'),
    path.join(__dirname, '../../../backend/policies/company-policy.md'),
    path.join(process.cwd(), 'backend/policies/company-policy.md'),
    path.join(process.cwd(), 'policies/company-policy.md'),
    path.resolve(process.cwd(), 'backend/policies/company-policy.md'),
    path.resolve(process.cwd(), 'policies/company-policy.md'),
  ];

  try {
    for (const policyPath of possiblePaths) {
      if (fs.existsSync(policyPath)) {
        const content = fs.readFileSync(policyPath, 'utf-8');
        console.log(`[Agent] Policy document loaded from: ${policyPath}`);
        return content;
      }
    }
    console.warn('[Agent] Policy document not found');
    return '';
  } catch (error: any) {
    console.error('[Agent] Error loading policy document:', error.message);
    return '';
  }
}

// Tool: Get policy context (rules)
function getPolicyContext(): string {
  try {
    const policyPath = path.join(__dirname, '../../policy.json');
    if (fs.existsSync(policyPath)) {
      const policy = JSON.parse(fs.readFileSync(policyPath, 'utf-8'));
      return JSON.stringify(policy.rules || [], null, 2);
    }
  } catch (error) {
    console.error('[Agent] Error loading policy:', error);
  }
  return 'No policy rules available';
}

// Tool: Get inbox stats
async function getInboxStats(): Promise<string> {
  try {
    const result = await pool.query(`
      SELECT COUNT(*) as total, 
             COUNT(*) FILTER (WHERE read_at IS NULL) as unread,
             COUNT(*) FILTER (WHERE EXISTS (
               SELECT 1 FROM contacts c WHERE c.id = messages.contact_id AND c.tags::text LIKE '%lead%'
             )) as leads
      FROM messages
    `);
    
    if (result.rows.length > 0) {
      const stats = result.rows[0];
      return `Inbox: ${stats.total} total messages, ${stats.unread} unread, ${stats.leads} leads.`;
    }
    return 'Inbox stats unavailable.';
  } catch (error: any) {
    return `Error fetching inbox stats: ${error.message}`;
  }
}

// Tool: Get tasks stats
async function getTasksStats(): Promise<string> {
  try {
    const result = await pool.query(`
      SELECT 
        COUNT(*) FILTER (WHERE priority = 'P0') as p0,
        COUNT(*) FILTER (WHERE priority = 'P1') as p1,
        COUNT(*) FILTER (WHERE priority = 'P2') as p2,
        COUNT(*) FILTER (WHERE status = 'pending') as pending
      FROM tasks
    `);
    
    if (result.rows.length > 0) {
      const stats = result.rows[0];
      return `Tasks: ${stats.pending} pending (${stats.p0} P0 urgent, ${stats.p1} P1 high, ${stats.p2} P2 normal).`;
    }
    return 'Tasks stats unavailable.';
  } catch (error: any) {
    return `Error fetching tasks stats: ${error.message}`;
  }
}

// LLM-based function calling: Determine which function to call
async function decideFunctionCall(question: string): Promise<{
  function: 'create_calendar_event' | 'get_policy_document' | 'answer_directly';
  confidence: number;
  reasoning: string;
}> {
  const systemPrompt = `You are a function router. Analyze the user's question and decide which function to call.

Available functions:
1. create_calendar_event - Call this if the user wants to schedule, book, plan, or create a calendar event, meeting, or appointment. Look for time/date references.
2. get_policy_document - Call this ONLY if the user explicitly mentions "policy", "policies", "rules", "guidelines", "escalation", or asks about company policies/rules.
3. answer_directly - Call this for all other questions (general chat, inbox questions, task questions, etc.)

Respond with ONLY a JSON object in this exact format:
{
  "function": "create_calendar_event" | "get_policy_document" | "answer_directly",
  "confidence": 0.0-1.0,
  "reasoning": "brief explanation"
}

Examples:
User: "Schedule a meeting tomorrow at 2pm"
→ {"function": "create_calendar_event", "confidence": 0.95, "reasoning": "User wants to schedule a meeting with specific time"}

User: "What is the refund policy?"
→ {"function": "get_policy_document", "confidence": 0.9, "reasoning": "User explicitly asks about policy"}

User: "Hi"
→ {"function": "answer_directly", "confidence": 1.0, "reasoning": "Simple greeting"}

User: "How many unread messages do I have?"
→ {"function": "answer_directly", "confidence": 1.0, "reasoning": "Question about inbox, not calendar or policy"}

Return ONLY the JSON, no other text.`;

  try {
    const llmResponse = await generateChatCompletion({
      model: process.env.LLM_MODEL || 'tinyllama',
      messages: [
        { role: 'system', content: systemPrompt },
        { role: 'user', content: question },
      ] as LLMMessage[],
      temperature: 0.1, // Low temperature for consistent function selection
      max_tokens: 150,
      useLocal: process.env.USE_OLLAMA !== 'false',
    });

    // Parse JSON response
    let decision;
    try {
      const jsonMatch = llmResponse.content.match(/\{[\s\S]*\}/);
      if (jsonMatch) {
        decision = JSON.parse(jsonMatch[0]);
      } else {
        decision = JSON.parse(llmResponse.content.trim());
      }
    } catch (parseError) {
      console.warn('[Agent] Failed to parse function decision, defaulting to answer_directly');
      console.warn('[Agent] LLM response:', llmResponse.content);
      return {
        function: 'answer_directly',
        confidence: 0.5,
        reasoning: 'Failed to parse LLM response',
      };
    }

    // Validate function name
    if (!['create_calendar_event', 'get_policy_document', 'answer_directly'].includes(decision.function)) {
      console.warn(`[Agent] Invalid function name: ${decision.function}, defaulting to answer_directly`);
      decision.function = 'answer_directly';
    }

    console.log(`[Agent] Function decision: ${decision.function} (confidence: ${decision.confidence})`);
    console.log(`[Agent] Reasoning: ${decision.reasoning}`);

    return {
      function: decision.function as 'create_calendar_event' | 'get_policy_document' | 'answer_directly',
      confidence: decision.confidence || 0.5,
      reasoning: decision.reasoning || 'No reasoning provided',
    };
  } catch (error: any) {
    console.error('[Agent] Error in function decision:', error);
    return {
      function: 'answer_directly',
      confidence: 0.5,
      reasoning: `Error: ${error.message}`,
    };
  }
}

// Main agentic chat function
export async function runAgenticChat(
  question: string,
  sessionId: string,
  conversationHistory?: Array<{ role: string; content: string }>
): Promise<{ answer: string; sessionId: string; calendarEvent?: any }> {
  try {
    // Remove question length limit (was 200 chars) - allow full questions
    // Only trim excessive whitespace
    const trimmedQuestion = question.trim();
    if (!trimmedQuestion) {
      throw new Error('Question cannot be empty');
    }
    // Get or create session state
    let state: AgentState;
    if (agentSessions.has(sessionId)) {
      state = agentSessions.get(sessionId)!;
      state.context.lastActivity = new Date();
    } else {
      state = {
        messages: [],
        sessionId,
        context: {
          lastActivity: new Date(),
        },
        tools: {},
      };
      agentSessions.set(sessionId, state);
    }

    // Load context if not already loaded
    if (!state.context.inboxStats || !state.context.tasksStats) {
      try {
        state.context.inboxStats = await getInboxStats();
        state.context.tasksStats = await getTasksStats();
      } catch (error) {
        console.warn('[Agent] Could not load context:', error);
      }
    }

    // Add conversation history to state
    if (conversationHistory && conversationHistory.length > 0) {
      const historyMessages = conversationHistory
        .slice(-10) // Last 10 messages
        .map((msg) => ({
          role: msg.role as 'user' | 'assistant',
          content: msg.content,
          timestamp: new Date(),
        }));
      state.messages.push(...historyMessages);
    }

    // Add current user message (no length limit - was 200 chars)
    state.messages.push({
      role: 'user',
      content: trimmedQuestion,
      timestamp: new Date(),
    });

    // Step 1: LLM decides which function to call
    const functionDecision = await decideFunctionCall(question);
    console.log(`[Agent] Decided to call: ${functionDecision.function} (confidence: ${functionDecision.confidence})`);

    let calendarEvent = null;
    let policyContext = '';
    let policyDocument = '';

    // Step 2: Execute the chosen function
    if (functionDecision.function === 'create_calendar_event' && functionDecision.confidence > 0.7) {
      console.log('[Agent] Executing: create_calendar_event (confidence:', functionDecision.confidence, ')');
      const result = await createCalendarEvent(question);
      console.log('[Agent] Calendar creation result:', result.success ? 'SUCCESS' : 'FAILED', result.message);
      
      if (result.success && result.event) {
        calendarEvent = result.event;
        state.tools.calendarEventCreated = result.event;
        
        // Format confirmation message
        const eventTime = new Date(calendarEvent.start_time).toLocaleString('en-US', {
          weekday: 'long',
          month: 'short',
          day: 'numeric',
          hour: 'numeric',
          minute: '2-digit',
        });
        
        const confirmationMessage = `✅ I've added "${calendarEvent.title}" to your calendar for ${eventTime}.${calendarEvent.location ? ` Location: ${calendarEvent.location}.` : ''}${calendarEvent.is_recurring ? ` This is a ${calendarEvent.recurrence_pattern} recurring event.` : ''}`;
        
        // Add confirmation to messages
        state.messages.push({
          role: 'assistant',
          content: confirmationMessage,
          timestamp: new Date(),
        });
        
        // Update session
        agentSessions.set(sessionId, state);
        
        console.log('[Agent] ✅ Returning calendar event confirmation');
        return {
          answer: confirmationMessage,
          sessionId,
          calendarEvent,
        };
      } else {
        // Calendar creation failed, fall through to answer with error
        console.warn('[Agent] Calendar event creation failed:', result.message);
        // Continue to generate a response explaining the failure
      }
    } else if (functionDecision.function === 'get_policy_document') {
      console.log('[Agent] Executing: get_policy_document');
      policyContext = getPolicyContext();
      policyDocument = getPolicyDocument();
      console.log(`[Agent] Loaded policy document (${policyDocument.length} chars)`);
    }
    // else: answer_directly - continue to LLM response

    // Step 3: Generate LLM response with appropriate context
    const isSimpleGreeting = /^(hi|hello|hey|greetings|good morning|good afternoon|good evening)$/i.test(question.trim());
    
    let systemPrompt: string;
    
    if (isSimpleGreeting) {
      // Minimal prompt for greetings
      systemPrompt = `You are Soraya AI. The user just said "${question}". Respond with a brief, friendly greeting. Do not list capabilities, context, policies, or any technical information. Keep it to 1-2 sentences maximum (under 50 words).`;
    } else {
      // Full prompt for other questions
      systemPrompt = `You are Soraya AI, a helpful assistant for managing inbox, tasks, and calendar.`;

      // Always add context if available - use it in your answers
      if (state.context.inboxStats || state.context.tasksStats) {
        systemPrompt += `\n\nIMPORTANT - Use this actual data when answering questions:\n${state.context.inboxStats || ''}\n${state.context.tasksStats || ''}\n\nWhen asked about inbox or tasks, use the numbers above. Don't say "I don't have access" - you have the data.`;
      }

      // Add policy information ONLY if get_policy_document was called
      if (functionDecision.function === 'get_policy_document' && policyDocument) {
        systemPrompt += `\n\nCompany Policy Document:\n${policyDocument}`;
        if (policyContext && policyContext !== 'No policy rules available') {
          try {
            const parsedRules = JSON.parse(policyContext);
            systemPrompt += `\n\nEscalation Rules:\n${JSON.stringify(parsedRules.slice(0, 5), null, 2)}`;
          } catch (e) {
            // Ignore parse errors
          }
        }
      }

      systemPrompt += `\n\nYour capabilities:
- Answer questions about inbox, tasks, and calendar using the context data provided above
- Create calendar events when asked (I will handle the creation automatically)

CRITICAL GUIDELINES:
- Use the actual context data provided above when answering questions
- Be natural and conversational
- Only provide information that's directly relevant to the question
- Don't mention policies, rules, or escalation keywords unless explicitly asked
- Keep responses SHORT and concise - maximum 150 words (preferably under 100 words)
- When user wants to schedule something, acknowledge that you'll add it to their calendar
- If you have context data, USE IT - don't say "I don't have access"`;

      if (calendarEvent) {
        systemPrompt += `\n\nNote: A calendar event was just created: ${(calendarEvent as any).title}`;
      }
    }

    // Build conversation messages for LLM
    // Include recent conversation history for context (increased from 8 to 15)
    const recentMessages = state.messages
      .slice(-15) // Last 15 messages for context (increased from 8)
      .map((msg) => ({
        role: msg.role,
        content: msg.content,
      }));
    
    // Smart truncation: prioritize policy document and context, truncate less important parts
    let finalSystemPrompt = systemPrompt;
    const maxSystemPromptLength = 3000; // Increased from 2000
    
    if (finalSystemPrompt.length > maxSystemPromptLength) {
      // If policy document is included, keep it and truncate other parts
      const policyDocMatch = finalSystemPrompt.match(/(Company Policy Document:[\s\S]+?)(?=\n\n|$)/);
      const contextMatch = finalSystemPrompt.match(/(IMPORTANT - Use this actual data:[\s\S]+?)(?=\n\n|$)/);
      
      // Keep essential parts
      let essentialParts = '';
      if (contextMatch) essentialParts += contextMatch[0] + '\n\n';
      if (policyDocMatch) essentialParts += policyDocMatch[0] + '\n\n';
      
      // Truncate capabilities/guidelines if needed
      const remainingLength = maxSystemPromptLength - essentialParts.length;
      const capabilitiesMatch = finalSystemPrompt.match(/(Your capabilities:[\s\S]+?)(?=\n\n|$)/);
      if (capabilitiesMatch) {
        const truncatedCapabilities = capabilitiesMatch[0].substring(0, Math.min(remainingLength - 200, capabilitiesMatch[0].length));
        finalSystemPrompt = essentialParts + truncatedCapabilities;
      } else {
        finalSystemPrompt = essentialParts + finalSystemPrompt.substring(0, remainingLength);
      }
      
      console.log(`[Agent] System prompt truncated from ${systemPrompt.length} to ${finalSystemPrompt.length} chars`);
    }
    
    // Check for truncated prompts and log warning
    if (finalSystemPrompt.length >= maxSystemPromptLength * 0.95) {
      console.warn(`[Agent] ⚠️  System prompt is near limit (${finalSystemPrompt.length}/${maxSystemPromptLength} chars). Some content may be truncated.`);
    }
    
    // Validate message structure before sending
    if (recentMessages.length === 0 && finalSystemPrompt.length === 0) {
      throw new Error('No messages to send to LLM');
    }
    
    const llmMessages: LLMMessage[] = [
      {
        role: 'system',
        content: finalSystemPrompt,
      },
      ...recentMessages,
    ];
    
    if (llmMessages[0].content.length === 0) {
      throw new Error('System prompt is empty');
    }

    // Generate response using LLM
    const useOllama = process.env.USE_OLLAMA !== 'false';
    const model = process.env.LLM_MODEL || 'tinyllama';
    
    const llmResponse = await generateChatCompletion({
      model,
      messages: llmMessages,
      temperature: 0.5, // Lower temperature for more focused responses
      max_tokens: 150, // Reduced from 200 to enforce brevity
      useLocal: useOllama,
    });

    if (!llmResponse || !llmResponse.content) {
      console.error('[Agent] ❌ LLM returned invalid response');
      throw new Error('LLM returned invalid response');
    }

    const answer = llmResponse.content.trim();
    
    // Log if answer seems truncated (very short or ends abruptly)
    if (answer.length < 10) {
      console.warn('[Agent] ⚠️  LLM response is unusually short, may be truncated');
    }

    // Add assistant response to state
    state.messages.push({
      role: 'assistant',
      content: answer,
      timestamp: new Date(),
    });

    // Keep only last 20 messages in state
    if (state.messages.length > 20) {
      state.messages = state.messages.slice(-20);
    }

    // Update session
    agentSessions.set(sessionId, state);

    return {
      answer,
      sessionId,
      calendarEvent,
    };
  } catch (error: any) {
    console.error('[Agent] Error running agentic chat:', error);
    throw error;
  }
}

// Get session state (for debugging)
export function getAgentSession(sessionId: string): AgentState | undefined {
  return agentSessions.get(sessionId);
}

// Clear session (for testing)
export function clearAgentSession(sessionId: string): void {
  agentSessions.delete(sessionId);
}

File: backend_python/app/agents/agentic_chat.py
"""Agentic AI Chat Agent with LLM-Based Function Calling."""
import os
import json
import asyncio
from typing import Dict, Any, List, Optional
from datetime import datetime, timedelta
from pathlib import Path
from app.db.connection import get_pool
from app.clients.llm import generate_chat_completion, LLMMessage, LLMRequestOptions
import httpx

# Agent state storage
agent_sessions: Dict[str, Dict[str, Any]] = {}


async def cleanup_inactive_sessions():
    """Clean up inactive sessions (older than 1 hour)."""
    now = datetime.now()
    one_hour = timedelta(hours=1)

    to_delete = []
    for session_id, state in agent_sessions.items():
        last_activity = state.get("context", {}).get("lastActivity")
        if last_activity and (now - last_activity) > one_hour:
            to_delete.append(session_id)

    for session_id in to_delete:
        del agent_sessions[session_id]
        print(f"[Agent] Cleared inactive session: {session_id}")


# Run cleanup every 5 minutes
async def start_cleanup_task():
    """Start background cleanup task."""
    while True:
        await asyncio.sleep(5 * 60)  # 5 minutes
        await cleanup_inactive_sessions()


# Cleanup task will be started in application lifespan
_cleanup_task = None


def get_cleanup_task():
    """Get the cleanup task instance."""
    return _cleanup_task


def set_cleanup_task(task):
    """Set the cleanup task instance."""
    global _cleanup_task
    _cleanup_task = task


async def create_calendar_event(text: str) -> Dict[str, Any]:
    """Create calendar event via API."""
    try:
        port = os.getenv("PORT", "3001")
        calendar_parse_url = f"http://localhost:{port}/api/calendar/parse"

        async with httpx.AsyncClient(timeout=30.0) as client:
            response = await client.post(
                calendar_parse_url,
                json={"text": text},
                headers={"Content-Type": "application/json"},
            )

            if response.status_code == 200:
                data = response.json()
                if data.get("event"):
                    return {
                        "success": True,
                        "event": data["event"],
                        "message": f'Calendar event "{data["event"]["title"]}" created successfully.',
                    }
            return {"success": False, "message": "Failed to create calendar event."}
    except Exception as error:
        print(f"[Agent] Calendar event creation error: {error}")
        return {"success": False, "message": f"Error: {str(error)}"}


def get_policy_document() -> str:
    """Get policy document content."""
    possible_paths = [
        Path(__file__).parent.parent.parent / "policies" / "company-policy.md",
        Path(__file__).parent.parent.parent.parent / "backend" / "policies" / "company-policy.md",
        Path.cwd() / "backend" / "policies" / "company-policy.md",
        Path.cwd() / "policies" / "company-policy.md",
    ]

    for policy_path in possible_paths:
        if policy_path.exists():
            try:
                content = policy_path.read_text(encoding="utf-8")
                print(f"[Agent] Policy document loaded from: {policy_path}")
                return content
            except Exception as error:
                print(f"[Agent] Error reading policy: {error}")

    return ""


def get_policy_context() -> str:
    """Get policy rules context."""
    try:
        policy_path = Path(__file__).parent.parent.parent / "policy.json"
        if not policy_path.exists():
            policy_path = Path(__file__).parent.parent.parent.parent / "backend" / "policy.json"

        if policy_path.exists():
            with open(policy_path, "r", encoding="utf-8") as f:
                policy = json.load(f)
                return json.dumps(policy.get("rules", []), indent=2)
    except Exception as error:
        print(f"[Agent] Error loading policy: {error}")

    return "No policy rules available"


async def decide_function_call(question: str) -> Dict[str, Any]:
    """Use LLM to decide which function to call."""
    system_prompt = """You are a function dispatcher. Decide which function to call based on the user's question.

Available functions:
1. create_calendar_event - Call this if the user wants to schedule, book, plan, or add a meeting/event/appointment
2. get_policy_document - Call this if the user asks about policies, rules, guidelines, procedures, or what to do
3. answer_directly - Call this for general questions, greetings, or when no specific function is needed

Return JSON with:
{
  "function": "create_calendar_event" | "get_policy_document" | "answer_directly",
  "confidence": 0.0-1.0,
  "reasoning": "brief explanation"
}"""

    try:
        llm_response = await generate_chat_completion(
            LLMRequestOptions(
                model=os.getenv("LLM_MODEL") or os.getenv("OPENAI_MODEL") or "tinyllama",
                messages=[
                    LLMMessage("system", system_prompt),
                    LLMMessage("user", question),
                ],
                temperature=0.3,
                max_tokens=100,
                use_local=os.getenv("USE_OLLAMA") != "false",
            )
        )

        # Parse JSON response
        content = llm_response.content.strip()
        content = content.replace("```json", "").replace("```", "").strip()
        try:
            decision = json.loads(content)
            return decision
        except json.JSONDecodeError:
            # Fallback: simple keyword matching
            question_lower = question.lower()
            if any(
                word in question_lower
                for word in ["schedule", "book", "plan", "meeting", "appointment", "event", "calendar", "add"]
            ):
                return {"function": "create_calendar_event", "confidence": 0.8, "reasoning": "Calendar-related keywords"}
            elif any(
                word in question_lower
                for word in ["policy", "rule", "guideline", "procedure", "what should", "how should"]
            ):
                return {"function": "get_policy_document", "confidence": 0.8, "reasoning": "Policy-related keywords"}
            else:
                return {"function": "answer_directly", "confidence": 0.9, "reasoning": "General question"}
    except Exception as error:
        print(f"[Agent] Error deciding function: {error}")
        return {"function": "answer_directly", "confidence": 0.5, "reasoning": "Error in decision"}


async def run_agentic_chat(
    question: str, session_id: str, conversation_history: List[Dict[str, str]]
) -> Dict[str, Any]:
    """Run agentic chat with function calling."""
    # Initialize or get session state
    if session_id not in agent_sessions:
        agent_sessions[session_id] = {
            "messages": [],
            "sessionId": session_id,
            "context": {"lastActivity": datetime.now()},
            "tools": {},
        }

    state = agent_sessions[session_id]
    state["context"]["lastActivity"] = datetime.now()

    # Fetch inbox and task stats
    inbox_stats = ""
    tasks_stats = ""
    try:
        pool = await get_pool()
        async with pool.acquire() as conn:
            inbox_row = await conn.fetchrow(
                """
                SELECT COUNT(*) as total, 
                       COUNT(*) FILTER (WHERE processed = false) as unread,
                       COUNT(*) FILTER (WHERE EXISTS (
                         SELECT 1 FROM contacts c WHERE c.id = messages.contact_id AND c.tags::text LIKE '%lead%'
                       )) as leads
                FROM messages
                """
            )
            if inbox_row:
                inbox_stats = f"Inbox: {inbox_row['total']} total, {inbox_row['unread']} unread, {inbox_row['leads']} leads."

            tasks_row = await conn.fetchrow(
                """
                SELECT 
                  COUNT(*) FILTER (WHERE priority = 'P0') as p0,
                  COUNT(*) FILTER (WHERE priority = 'P1') as p1,
                  COUNT(*) FILTER (WHERE status = 'pending') as pending
                FROM tasks
                """
            )
            if tasks_row:
                tasks_stats = f"Tasks: {tasks_row['pending']} pending ({tasks_row['p0']} P0 urgent, {tasks_row['p1']} P1 high)."
    except Exception as error:
        print(f"[Agent] Error fetching context: {error}")

    # Check if simple greeting
    is_simple_greeting = question.strip().lower() in [
        "hi",
        "hello",
        "hey",
        "greetings",
        "good morning",
        "good afternoon",
        "good evening",
    ]

    if is_simple_greeting:
        # Simple greeting response
        llm_response = await generate_chat_completion(
            LLMRequestOptions(
                model=os.getenv("LLM_MODEL") or os.getenv("OPENAI_MODEL") or "tinyllama",
                messages=[
                    LLMMessage(
                        "system",
                        "You are a helpful assistant. Respond to greetings naturally and briefly. Keep it to 1-2 sentences maximum (under 50 words).",
                    ),
                    LLMMessage("user", question),
                ],
                temperature=0.5,
                max_tokens=50,
                use_local=os.getenv("USE_OLLAMA") != "false",
            )
        )
        return {
            "answer": llm_response.content.strip(),
            "sessionId": session_id,
        }

    # Decide which function to call
    function_decision = await decide_function_call(question)

    calendar_event = None

    # Execute function based on decision
    if function_decision["function"] == "create_calendar_event" and function_decision.get("confidence", 0) > 0.7:
        # Create calendar event
        calendar_result = await create_calendar_event(question)
        if calendar_result.get("success"):
            calendar_event = calendar_result["event"]
            state["tools"]["calendarEventCreated"] = calendar_event

    # Build system prompt
    policy_rules = ""
    policy_document = ""
    if function_decision["function"] == "get_policy_document":
        policy_rules = get_policy_context()
        policy_document = get_policy_document()

    # Build recent messages from conversation history
    recent_messages = []
    if conversation_history:
        recent_messages = [
            LLMMessage(msg["role"], msg["content"])
            for msg in conversation_history[-15:]
            if msg.get("role") in ["user", "assistant"]
        ]
    elif state["messages"]:
        recent_messages = [
            LLMMessage(msg["role"], msg["content"])
            for msg in state["messages"][-15:]
            if msg.get("role") in ["user", "assistant"]
        ]

    # Build system prompt
    system_prompt_parts = [
        "You are a helpful AI assistant for managing inbox, tasks, and calendar.",
        "You can help create calendar events, answer questions about policies, and provide insights about the inbox.",
    ]

    if inbox_stats or tasks_stats:
        system_prompt_parts.append(f"\nCurrent context:\n{inbox_stats}\n{tasks_stats}")

    if policy_rules and policy_document:
        system_prompt_parts.append(f"\nPolicy Rules:\n{policy_rules}")
        system_prompt_parts.append(f"\nPolicy Document:\n{policy_document[:2000]}")  # Limit length

    system_prompt_parts.append(
        "\nInstructions:\n- Use the context provided to answer questions accurately.\n- Be concise and direct. Maximum 250 words.\n- If you created a calendar event, mention it naturally in your response."
    )

    system_prompt = "\n".join(system_prompt_parts)

    # Truncate if too long
    if len(system_prompt) > 3000:
        system_prompt = system_prompt[:3000] + "\n[Context truncated...]"
        print("[Agent] Warning: System prompt truncated")

    # Generate response
    messages = [LLMMessage("system", system_prompt)]
    messages.extend(recent_messages)
    messages.append(LLMMessage("user", question))

    try:
        llm_response = await generate_chat_completion(
            LLMRequestOptions(
                model=os.getenv("LLM_MODEL") or os.getenv("OPENAI_MODEL") or "tinyllama",
                messages=messages,
                temperature=0.5,
                max_tokens=150,
                use_local=os.getenv("USE_OLLAMA") != "false",
            )
        )

        answer = llm_response.content.strip()

        # Add calendar event confirmation if created
        if calendar_event:
            answer += f'\n\n✅ Calendar event "{calendar_event.get("title")}" has been added to your calendar.'

        # Update session state
        state["messages"].append({"role": "user", "content": question, "timestamp": datetime.now()})
        state["messages"].append({"role": "assistant", "content": answer, "timestamp": datetime.now()})

        return {
            "answer": answer,
            "sessionId": session_id,
            "calendarEvent": calendar_event,
        }
    except Exception as error:
        print(f"[Agent] Error generating response: {error}")
        return {
            "answer": "I'm having trouble processing that request. Please try again.",
            "sessionId": session_id,
        }

